---
title: 'Biology 351: Lab Week 10 S23'
output:
  pdf_document: default
  html_notebook: default
---

*Introduction*

We are back to R to analyze and plot our allele frequency data for the last several weeks of class. Before spring break, you generated a filtered vcf file using just your two samples; you'll use this next week to do an FST analysis looking at your two populations. Over spring break, I created the same filtered vcf file that includes all of the samples--34 different pooled DNA samples from the past 6 years in Virginia and from all over the east coast in Fall 2022. This week we are going to use these data to run and  a principal component analysis and do a preliminary comparison of FST across all the samples before moving on to our focal samples. 

To start, I have already turned the data encoded in the AD and RD spaces of the vcf into an R data table.

**Question:** (From a few weeks ago): What do AD and RD stand for? What information do they hold?
**Answer here:**

**Question** How would you convert RD and AD values into an allele frequency? Note that we are calculating frequencies of the reference allele. 
**Answer here**

We need to use some bash code to get some files off of `spydur` and to the Rstudio server.  We will use a Linux function called `scp` to do that. It will copy the named files to your current working directory on the Rstudio server (indicated by the `./` at the end of the command). Luckily for us, we can run Linux commands through Rstudio. 

Replace `perickso` with your netid below and run this bash code chunk.  You should see these files appear in your working folder to the right. The first line sets up some information that will allow us to use one of the R packages that we'll be using today. 

Note these files will take a ~30 seconds to download.

```{bash}
export LD_LIBRARY_PATH=/usr/lib64:/usr/local/sw/anaconda3/lib:/usr/pgsql-13/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/sw/anaconda3/lib
scp perickso@spydur:~/shared_perickso/ZP_chr1.recode.vcf ./
scp perickso@spydur:~/shared_perickso/ZP_chr3.recode.vcf ./
scp perickso@spydur:~/shared_perickso/downsampled_allele_freqs_S23.Rdata ./
scp perickso@spydur:~/shared_perickso/zap_pool_info_W22_forclass.csv ./
scp perickso@spydur:~/shared_perickso/vcf_sample_names.txt ./
```

*Part 1: Principal Components Analysis (PCA)*

First, let's load the packages and data in. I've already been processed the allele frequencies into a data.table for each pooled sample, but it's a big file and will take ~30 seconds to load. It will load in as an object called `y`.
 
```{r}
library(data.table)
library(foreach)
library(ggplot2)

#read in vcf allele frequency data. Will load in as a data.table called freqs.wide
load("downsampled_allele_freqs_S23.Rdata")
```


When we do our analysis we want to remove SNPs  that are closely linked. These SNPs will tend to have correlated allele frequencies and may bias our analysis. I ran a short program that will keep one SNP in every 2000 bp window of the genome. That way, any two SNPs in the analysis are, on average, about 2kb apart, which is far enough to remove most linked SNPs. 

**Question** Look at the `y` table. How are the data organized?

**Answer here:**

**Question** How many SNPs are currently in our dataset? 

**Answer here:**

First let's do a principal components analysis using our  data for the entire genome. Note that while most of the PCAs we have looked at in class have used *genotypes of individuals* at many SNPs to calculate PCA, we can also use *allele frequencies of populations* for a similar calculation. 

Now we'll do the actual principal components calculation. R has a built-in function called `prcomp` to do this. You can run the whole chunk of code below to execute the PCA.
```{r}
#we only want to use SNPs with no missing data. This function will remove any rows of the matrix with missing data
pca.matrix<-as.matrix(y[,3:37])
#this assigns marker names to each row of the matrix
  rownames(pca.matrix)=y$marker  
  #omit missing data
pca.matrix.complete<-na.omit(pca.matrix)
  #calculate PCA with prcomp() function
  pca.results<-prcomp(pca.matrix.complete, scale=T)
  #turn PCA results into a data.table
  pca.data<-as.data.table(pca.results$rotation)
  #add ID columns to that table
  pca.data[,Pool:=rownames(pca.results$rotation)]
  #look at just PC1 through PC4
  pca.data<-pca.data[,.(Pool, PC1, PC2, PC3, PC4)]

```
**Question** Look at the `pca.data` object. What does it have in the rows and columns? 

**Answer here**

**Question** Based on what you know about principle components and how these samples were generated and processed,  what does it mean if two of our samples samples have similar values of PC1 and PC2?

**Answer here**

To make an informative plot, we need to bring in some information about each sample so we know what we are looking at and can color-code our points appropriately. Update the code below to merge the sample information into the PCA data. 

```{r}
info<-fread("zap_pool_info_W22_forclass.csv")
```

**Question** Look at the `info` table you just loaded. What is the name of the column that is shared in common with "pca.data"?

**Answer here**

Now set up the command below to merge the sample and pca data by replacing `"name.of.shared.column"` with the column you identified. Keep the quotes around the column name. Note that this command takes the names of two data tables and the column they have in common, then uses a shared column to line up the data. 

```{r}
pca.data<-merge(pca.data, info, by="name.of.shared.column") 
```

**Question** Now look at your pca.data table. What did the merge function just do?
**Answer here**

Now let's make some plots of our whole-genome PCA data. First make a plot comparing the samples by collection location by completing the x, y, and color information below to make a standard PCA plot.

```{r}
ggplot(pca.data)+geom_point(aes(x=   , 
                            y=  , 
                            color=    )) 
```

**Question** Do you see any patterns  based on the location of the samples? 
**Answer here**

You can also plot only some of your data by using conditional statements to extract only some rows. Note how this plot uses a conditional statement to only plot the data collected in 2021. Recall that conditional statements require two  == signs. 

```{r}
ggplot(pca.data[Year==2021])+geom_point(aes(x= PC1, 
                            y= PC2, 
                            color= Location)) 
```
**Question** Do the samples appear to cluster by location when looking at only one year?
**Answer here**


Next, let's make a plot that highlights the two samples you'll be comparing for your analysis. We will first need to make a new column that identifies the samples you worked on. You'll need this plot for your final presentation so make sure you get this part done correctly.

```{r}
#first make a new column called "focal_sample" that will be filled with "FALSE" values (every row will say "FALSE" for this column). You don't need to change anything about this line of code
pca.data[, focal_sample:=FALSE]

#now, for the rows that contain your two samples, change the focal_sample column to TRUE using an == statement to replace conditional.statement.here before the comma. Hint: your conditional statement should start with Pool
pca.data[ conditional.statement.here , focal_sample:=TRUE] #complete this line
pca.data[ conditional.statement.here , focal_sample:=TRUE] #complete this line

#print pca.data to make sure this worked; if not ask me for help
pca.data
```

Now you can use your new column to make a plot to identify your samples in the whole dataset:

```{r}
ggplot(pca.data)+geom_point(aes(x=  , 
                            y=   ,
                            color=  ))
```

**Question** Compared to the overall variation in the samples, how similar are your two samples that you'll be comparing? Do they differ in PC1, PC2, or both?
**Answer here**

We were focusing on the whole genome, but previous research on Zaprionus has suggested that there are some interesting trends of population differentiation with individual chromosomes. Let's repeat our PCA but now we will calculate the PCA for each chromosome separately and compare. Like a few weeks ago, we'll use a "for loop" that processes the chromosomes one at a time. The mechansims of how this code works are too complicated to go through today, so you can just run this code and proceed to the steps below. 

```{r}
pca.chrom<-foreach(chrom=c(1:5))%do%{
  print(chrom)
  #this line will pull out only the data for one chromosome
  y.chrom<-y[chr==chrom,3:37]
  pca.matrix.chrom<-as.matrix(y.chrom)
#this assigns marker names to each row of the matrix
  rownames(pca.matrix.chrom)=y.chrom$marker
#only want SNPs with calls in all pops
  pca.matrix.chrom.complete<-na.omit(pca.matrix.chrom)
  #calculate PCA for the chromosome
  pca.results.chrom<-prcomp(pca.matrix.chrom.complete, scale=T)
  #turn it into a data table
  pca.data.chrom<-as.data.table(pca.results.chrom$rotation)
  #add names of pools
  pca.data.chrom[,Pool:=rownames(pca.results.chrom$rotation)]
  pca.data.chrom[,chrom:=chrom]
  return(pca.data.chrom)
}
pca.chrom<-rbindlist(pca.chrom)
```
**Question** Which line of the code above caused the analysis to be restricted to just one chromosome at a time?

**Answer here**

Now use the merge command that you learned above to merge your new `pca.chrom` object with the sample information. Make sure you change all the variables that need to be changed to execute this command.

```{r}
#merge here using similar command to above (or see cheat sheet)


```
Confirm that your merging worked as intended before proceeding

Now we'll make the same plots we made above, but plot each chromosome separately with the `facet_wrap` command in `ggplot`. To add facets you add `+ facet_wrap(~variable.to.facet.by, scales="free")` to the end of your plotting command. Use your code from above as an example to make a new plot that is color-coded by Location and divided by chromosome so that there is a different plot for each chromosome. 

```{r}
#write plotting code here




```


**Question** How do the individual chromosome results compare to the full genome results? What could explain the differences? Do any chromosomes stand out? 

**Answer here**

*Part 2: Independent investigation*

You've now seen how to make PCA plots that color coded, and you've seen how to subset data (use only some of the data available with conditional statements, like [Year==2021]) and how to split your data into multiple graphs with facet_wrap(). Your challenge now is to use the various columns available in the sample information to come up with another way to plot the data to look for patterns. You could look at a single orchard over time, you could compare seasons, you could compare multiple orchards in one year,  you could look at multiple locations within a year, etc. Below are the metadata describing the different columns to help you.

    Pool: The name of the sequencing library.
    Location: The location flies were collected:
        CM = Carter Mountain, Charlottesville, VA
        FL = Fruit and Spice Park, Miami, FL
        HPO = Hanover Peach Orchard, Richmond, VA
        GA = Hillcrest Orchard, Ellijay, GA
        CK = Cross Keys Orchard, Harrisonburg, VA
        MA = Acton, MA (flies combined from multiple orchards)
        CT = Lyman Orchards, Middleton, CT
        PA = Linvilla Orchards, Media, PA
        CH = Chiles Orchards, Crozet, VA
    Experiment: Whether the sample was part of the CM seasonal sampling, HPO seasonal sampling, or latitudeinal experiment
    Experiment 2: Identifies all samples collected in Virginia in 2002 from 4 different orchards
    State: The location the samples were collected
    Year: The year of sampling
    Month: The calendar month of sampling
    Season: whether the sample was collected early (July-August), middle (September), or late (October-November) during hte Z. indianus growing season
    Sample Date: The day flies were collected
    females: The number of female flies in the pool
    males: The number of male flies in the pool
    total: The total number of flies in the pool

**Question** What is the PCA pattern you are interested in looking at and why?
**Answer**

**Question** How will you set up your analysis? What data will you use, how will you color code, and will you facet your graph?
**Answer**

Before moving on, check with Dr. E about your answers. Then, develop the code for your plot below and make your plot. 
```{r}



```

At the end of the lab you will be asked to write a brief figure legend and results statement for this analysis, but for now keep coding! 

*Part 3: Genome-wide FST between all pairs of samples*

For this part we are going to use a package called `poolfstat`. This package is specifically designed to do FST and other population genetic tests on pooled sequencing data like we have. 

Run the code below to read our full vcf file into R with poolfstat. It needs to know the names and number of individuals in each pool, which is found in our previously-loaded "info" file. Because the order of the samples in the vcf is is different than the order in the sample file, we have to tell it to read the poolnames and poolsizes out of order. This step will take a couple minutes but will update you on its progress.

To speed things up, we will just be conducting this analysis on a small region of the genome on chromosome 3.

```{r}
library(poolfstat)

sample.names=fread("vcf_sample_names.txt", header =F)$V1
setkey(info, Pool)
pool.sizes=info[sample.names, total]
vcf.dat<-vcf2pooldata(vcf.file="ZP_chr3.recode.vcf",
                  poolnames=sample.names,
                  poolsizes=pool.sizes,
                  min.cov.per.pool=25)
```
One of the functions of this package can calculate "pairwise" FST using each sample. 

**Question** What do higher FST values mean?
**Answer here**

**Question:** What do you think "pairwise' FST means? 
**Answer here**

The function is called `compute.pairwiseFST` and it takes our imported vcf file as an input. Write the command below to do the calculation and save the output as an object called `pfst`. Now write your command below and run it. It takes a couple minutes to run.

```{r}
pfst<-compute.pairwiseFST(vcf.dat,verbose=TRUE) 
```

Let's look at the `pfst` object created in our last command. It created a bunch of data, but the actual FST values are saved as something called `PairwiseFSTmatrix`, the first 10 rows and columns of which will be printed with the command below

```{r}
pfst@PairwiseFSTmatrix[1:10, 1:10]
```

**Question** What is shown in these data?
**Answer here**

**Question** Why are there "NA"s along a diagonal line?
**Answer here**

The poolfstat package includes a built-in function to create a heatmap of the FST values. Heatmaps use different colors to show numerical data.

```{r}
heatmap(pfst,symm=T)

```

**Question** What are some things that are wrong with this plot? You should be able to identify at least 2-3 issues.
**Answer here**



Let's make a better heatmap so that we can be certain of what we are looking at. You'll need to run the first three lines, then look at the new object you've created and update the plotting information in the second chunk of code to actually  make the plot.

```{r}
#first we melt our data into a long format
library(reshape2)
library(viridis)
fst.melt<-as.data.table(melt(pfst@PairwiseFSTmatrix))
names(fst.melt)=c("Sample1", "Sample2", "FST")

```

Now look at the FST.melt object. You'll need to understand how it's laid out to update the code for your heatmap below.

```{r}
#now update the code to make a plot
ggplot(fst.melt)+
  geom_tile(aes(x=  , #what column makes the x axis? 
                y=  , #what column makes the y axis?  
                fill=   ))+ #what column should determine the coloring?  
  #this next line is going to rotate the angle of the letters on teh x axis
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
  #this line removes x and y axis labels and labels the color scale
  labs(x=NULL, y=NULL, fill="FST")+
  #this line will use prettier colors!
  scale_fill_viridis()
```
**Question** Which samples appear to be the most different from all the other samples? Based on the `info` table, what do they have in common? 
**Answer here**

**Question** does this result support or contradict the chromosome-specific PCA analysis?
**Answer here**

Now we are going to repeat everythign we just did but for a different, randomly chosen region of the genome (somewhere in the middle of chromosome 1) for comparison's sake. Run the whole chunk of code below, you don't have to change anything. It will take a few seconds to run.
```{r}
vcf.dat<-vcf2pooldata(vcf.file="ZP_chr1.recode.vcf",
                  poolnames=sample.names,
                  poolsizes=pool.sizes,
                  min.cov.per.pool=25)

pfst<-compute.pairwiseFST(vcf.dat,verbose=TRUE) 
fst.melt<-melt(pfst@PairwiseFSTmatrix)
ggplot(fst.melt)+
  geom_tile(aes(x= Var1 , 
                y= Var2 , 
                fill= value  ))+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
  labs(x=NULL, y=NULL, fill="FST")+
  #this line will use prettier colors!
  scale_fill_viridis()
```
**Question** Looking at the FST scale, do there seem to be a lot of differences between samples in this region of the genome?  
**Answer here**

**Question** Find the point that is the intersection of your two samples--How does the FST between your two samples compare to the range of values in all the comparisons? Relatively low? High? 
**Answer here**

**Final task**For the independent analysis that you conducted in part 2, write a figure legend and a short results statement describing the results and citing the figure. Remember that your figure legend should allow the reader to understand what was done to produce the data. The results paragraph should orient the reader to purpose of the experiment/analysis and then describe key results, referring back to the figure (which you can name "Figure 1")

**Figure legend here**

**Results here**

When you are done, please use Preview-->Knit to HTML to generate a nice preview, then save it as a PDF and upload to Blackboard. Make sure the PDF includes all of your plots! 